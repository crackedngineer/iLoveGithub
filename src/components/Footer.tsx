import React from "react";
import { Heart } from "lucide-react";

const Footer = () => {
  const currentYear = new Date().getFullYear();

  return (
    <footer className="w-full py-6 px-6 bg-github-lightGray dark:bg-github-darkBlue border-t border-gray-200 dark:border-gray-800">
      <div className="flex flex-col md:flex-row justify-between items-center text-sm text-gray-500 dark:text-gray-400">
        {/* Copyright Section */}
        <p>&copy; {currentYear} I Love GitHub. All rights reserved.</p>

        {/* Creator Section */}
        <p className="flex items-center gap-1 mt-3 md:mt-0">
          Made with <Heart className="h-4 w-4 text-red-500" /> by{" "}
          <a
            href="https://github.com/subhomoy-roy-choudhury/iLoveGithub"
            target="_blank"
            rel="noopener noreferrer"
            className="hover:text-github-blue dark:hover:text-white transition-colors"
          >
            I Love GitHub Team
          </a>
        </p>

        {/* Sitemap Link */}
        <a
          href="/sitemap.xml"
          className="mt-3 md:mt-0 hover:text-github-blue dark:hover:text-white transition-colors"
        >
          Sitemap
        </a>
      </div>
    </footer>
  );
};

export default Footer;
