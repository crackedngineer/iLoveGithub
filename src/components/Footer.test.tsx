import {render, screen} from "@testing-library/react";
import Footer from "./Footer";

// Mock next/image
jest.mock("next/image", () => ({
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line jsx-a11y/alt-text
    return <img {...props} />;
  },
}));

describe("Footer Component", () => {
  beforeEach(() => {
    // Mock Date to return consistent year for testing
    jest.useFakeTimers();
    jest.setSystemTime(new Date("2024-01-01"));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  describe("Rendering", () => {
    it("renders the footer element", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer).toBeInTheDocument();
    });

    it("displays copyright text with current year", () => {
      render(<Footer />);

      expect(screen.getByText(/© 2024 iLoveGithub/i)).toBeInTheDocument();
    });

    it('displays "Made in" text', () => {
      render(<Footer />);

      expect(screen.getByText("Made in")).toBeInTheDocument();
    });

    it('displays "by" text', () => {
      render(<Footer />);

      expect(screen.getByText("by")).toBeInTheDocument();
    });

    it("displays creator name as link", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link", {name: /subhomoy roy choudhury/i});
      expect(creatorLink).toBeInTheDocument();
    });

    it("renders Indian flag image", () => {
      render(<Footer />);

      const flagImage = screen.getByAltText("Indian Flag");
      expect(flagImage).toBeInTheDocument();
    });
  });

  describe("Links", () => {
    it("has correct href for creator GitHub profile", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link", {name: /subhomoy roy choudhury/i});
      expect(creatorLink).toHaveAttribute("href", "https://github.com/crackedngineer/");
    });

    it("opens creator link in new tab", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link", {name: /subhomoy roy choudhury/i});
      expect(creatorLink).toHaveAttribute("target", "_blank");
    });

    it("has security attributes on external link", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link", {name: /subhomoy roy choudhury/i});
      expect(creatorLink).toHaveAttribute("rel", "noopener noreferrer");
    });

    it("has hover effect classes on creator link", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link", {name: /subhomoy roy choudhury/i});
      expect(creatorLink).toHaveClass(
        "hover:text-github-blue",
        "dark:hover:text-white",
        "transition-colors",
      );
    });
  });

  describe("Image", () => {
    it("renders flag image with correct src", () => {
      render(<Footer />);

      const flagImage = screen.getByAltText("Indian Flag");
      expect(flagImage).toHaveAttribute("src", "/icons/indian-flag.svg");
    });

    it("renders flag image with correct dimensions", () => {
      render(<Footer />);

      const flagImage = screen.getByAltText("Indian Flag");
      expect(flagImage).toHaveAttribute("width", "18");
      expect(flagImage).toHaveAttribute("height", "18");
    });

    it("has alt text for flag image", () => {
      render(<Footer />);

      const flagImage = screen.getByAltText("Indian Flag");
      expect(flagImage).toBeInTheDocument();
    });
  });

  describe("Styling", () => {
    it("applies correct footer styling classes", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer).toHaveClass(
        "w-full",
        "py-6",
        "px-4",
        "sm:px-6",
        "bg-github-light-gray",
        "dark:bg-github-dark-blue",
        "border-t",
        "border-gray-200",
        "dark:border-gray-800",
      );
    });

    it("applies responsive flexbox layout", () => {
      const {container} = render(<Footer />);
      const contentDiv = container.querySelector(".flex.flex-col");

      expect(contentDiv).toHaveClass("md:flex-row", "md:justify-center", "md:items-center");
    });

    it("applies text styling classes", () => {
      const {container} = render(<Footer />);
      const contentDiv = container.querySelector(".text-sm");

      expect(contentDiv).toHaveClass(
        "text-center",
        "md:text-left",
        "text-gray-500",
        "dark:text-gray-400",
      );
    });

    it("hides separator on mobile screens", () => {
      const {container} = render(<Footer />);
      const separator = container.querySelector(".hidden.sm\\:inline");

      expect(separator).toBeInTheDocument();
      expect(separator).toHaveTextContent("•");
    });
  });

  describe("Layout", () => {
    it("displays content in correct order", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");
      const text = footer?.textContent || "";

      // Check order of elements
      expect(text).toMatch(/© 2024 iLoveGithub.*•.*Made in.*by.*Subhomoy Roy Choudhury/);
    });

    it("groups creator info together", () => {
      render(<Footer />);

      // All creator info should be within the same paragraph
      const creatorParagraph = screen.getByText("Made in").closest("p");
      expect(creatorParagraph).toBeInTheDocument();
      expect(creatorParagraph?.textContent).toContain("Made in");
      expect(creatorParagraph?.textContent).toContain("by");
      expect(creatorParagraph?.textContent).toContain("Subhomoy Roy Choudhury");
    });
  });

  describe("Responsive Design", () => {
    it("applies mobile-first padding", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer).toHaveClass("px-4", "sm:px-6");
    });

    it("centers content on mobile", () => {
      const {container} = render(<Footer />);
      const contentDiv = container.querySelector(".text-center");

      expect(contentDiv).toBeInTheDocument();
    });

    it("applies desktop-specific layout", () => {
      const {container} = render(<Footer />);
      const contentDiv = container.querySelector(".flex");

      expect(contentDiv).toHaveClass("md:flex-row", "md:justify-center");
    });
  });

  describe("Accessibility", () => {
    it("has semantic footer element", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer?.tagName.toLowerCase()).toBe("footer");
    });

    it("has descriptive alt text for flag image", () => {
      render(<Footer />);

      const flagImage = screen.getByAltText("Indian Flag");
      expect(flagImage).toHaveAttribute("alt", "Indian Flag");
    });

    it("has accessible link text", () => {
      render(<Footer />);

      const creatorLink = screen.getByRole("link");
      expect(creatorLink).toHaveAccessibleName(/subhomoy roy choudhury/i);
    });
  });

  describe("Dynamic Year", () => {
    it("displays copyright year from Date object", () => {
      render(<Footer />);

      // The year should be 2024 based on our mock
      expect(screen.getByText(/© 2024 iLoveGithub/i)).toBeInTheDocument();
    });

    it("calculates year dynamically on render", () => {
      // Restore original Date
      jest.restoreAllMocks();

      render(<Footer />);

      const currentYear = new Date().getFullYear();
      const copyrightText = screen.getByText(new RegExp(`© ${currentYear} iLoveGithub`, "i"));
      expect(copyrightText).toBeInTheDocument();
    });
  });

  describe("Dark Mode Support", () => {
    it("has dark mode background classes", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer).toHaveClass("dark:bg-github-dark-blue");
    });

    it("has dark mode text color classes", () => {
      const {container} = render(<Footer />);
      const contentDiv = container.querySelector(".text-gray-500");

      expect(contentDiv).toHaveClass("dark:text-gray-400");
    });

    it("has dark mode border classes", () => {
      const {container} = render(<Footer />);
      const footer = container.querySelector("footer");

      expect(footer).toHaveClass("dark:border-gray-800");
    });
  });

  describe("Edge Cases", () => {
    it("renders without crashing", () => {
      expect(() => render(<Footer />)).not.toThrow();
    });

    it("handles missing year gracefully", () => {
      // Even with date mocked, component should render
      render(<Footer />);

      const footer = screen.getByRole("contentinfo");
      expect(footer).toBeInTheDocument();
    });
  });

  describe("Snapshot", () => {
    it("matches snapshot", () => {
      const {container} = render(<Footer />);
      expect(container).toMatchSnapshot();
    });
  });
});
