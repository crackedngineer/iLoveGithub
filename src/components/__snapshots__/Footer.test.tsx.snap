// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Footer Component Snapshot matches snapshot 1`] = `
<div>
  <footer
    class="w-full py-6 px-4 sm:px-6 bg-github-light-gray dark:bg-github-dark-blue border-t border-gray-200 dark:border-gray-800"
  >
    <div
      class="flex flex-col gap-2 md:flex-row md:justify-center md:items-center text-sm text-center md:text-left text-gray-500 dark:text-gray-400"
    >
      <p>
        © 
        2024
         iLoveGithub
      </p>
      <span
        class="hidden sm:inline"
      >
        •
      </span>
      <p
        class="flex justify-center md:justify-start items-center gap-2"
      >
        <span>
          Made in
        </span>
        <img
          alt="Indian Flag"
          height="18"
          src="/icons/indian-flag.svg"
          width="18"
        />
        <span>
          by
        </span>
        <a
          class="hover:text-github-blue dark:hover:text-white transition-colors"
          href="https://github.com/crackedngineer/"
          rel="noopener noreferrer"
          target="_blank"
        >
          Subhomoy Roy Choudhury
        </a>
      </p>
    </div>
  </footer>
</div>
`;
